docs/proposals/20220927-label-sync-between-machine-and-nodes.md

```diff
@@ -0,0 +1,131 @@
+
+#### Synchronization of CAPI Labels
+
+Only the labels that fall under the prefix described above will need to be kept in sync between the Machine and Node objects.
+
+The syncronization process is going to use [server side apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/) in order to ensure that Cluster API will manage only the subset of labels coming from Machine objects and ignore labels applied concurrently by other users/other components.
+
+This requires to define an identity/manager name to be used by CAPI when performing this operation; additionally during implementation we are going to identify and address eventual additional steps required to properly transition existing Nodes to SSA, if required.
+
+The syncronization process will be implemented in the Machine controller, so reconciliation is triggered both when Machine object changes or when the Node changes (the machine controller watches Nodes within the workload cluster).
+
+#### Delay between Node Create and Label Sync
+
+The Node object is first created when kubeadm joins a node to the workload cluster (i.e. kubelet is up and running). There may be a delay (potentially several seconds) before the machine controller kicks in to apply the labels on the Node.
+
+Kubernetes supports both equality and inequality requirements in label selection. In an equality based selection, the user wants to place a workload on node(s) matching a specific label (e.g. Node.Labels contains `my.prefix/foo=bar`). The delay in applying the label on the node, may cause a subsequent delay in the placement of the workload, but this is likely acceptable.
+
+In an inequality based selection, the user wants to place a workload on node(s) that do not contain a specific label (e.g. Node.Labels not contain `my.prefix/foo=bar`). The case is potentially problematic because it relies on the absence of a label and this can occur if the pod scheduler runs during the delay interval.
+
+One way to address this is to use kubelet's `--register-with-taints` flag. Newly minted nodes can be tainted via the taint `cluster.x-k8s.io=label-sync-pending:NoSchedule`. Assuming workloads don't have this specific toleration, then nothing should be scheduled. KubeadmConfigTemplate provides the means to set taints on nodes (see  JoinConfiguration.NodeRegistration.Taints).
+
+The process of tainting the nodes, can be carried out by the user and can be documented as follows:
+
+```
+If you utilize inequality based selection for workload placement, to prevent unintended scheduling of pods during the initial node startup phase, it is recommend that you specify the following taint in your KubeadmConfigTemplate:
+`cluster.x-k8s.io=label-sync-pending:NoSchedule`
+```
+
+After the node has come up and the machine controller has applied the labels, the machine controller will also remove this specific taint if it's present.
+
+In the future, we may consider automating the insertion of the taint via CABPK.
```

Comment:
```suggestion
```

Reply:
We should do this now if that's the preferred path?

Reply:
@enxebre opinions?

Pushing this a little bit further, we can even also add this in the bootstrap provider contracts as an optional (but recommended) behavior to support inequality-based selection for workload placement

Reply:
My only concern is that built-in tainting would has a workload scheduling delay impact on every upcoming Node, may be that's negligible. May be is not a bad idea start with letting it be opt-in and revisit based on usage feedback.

Reply:
Isn't the delay wanted in this case? If let's say a user is scaling up a cluster to satisfy workload requirements, and we have a number of pods waiting to be scheduled on new nodes, if there are affinity rules, it might be the case that pods gets scheduled on the wrong node to begin with.

Reply:
yes, but if built-in the delay would also apply unnecessarily to scenarios where no affinity is intended at all, but may be that's just fine.

Reply:
I wonder how big the delay is. IIRC usually the cloud-provider has a taint as well which is only removed after the providerID is set. I can imagine that it doesn't take that long from the cloud-provider setting the providerID/dropping their taint to us removing our taint

Reply:
I don't think that the delay will change from users adding the taint in the kubeadmConfig as we are suggesting above to CAPI Adding the taint by default.
The UX will improve though, so I will rephrase this so we explore this option

Reply:
> I don't think that the delay will change from users adding the taint in the kubeadmConfig as we are suggesting above to CAPI Adding the taint by default.

I think Alberto's point was the delay between not having the taint at all and automatically having it via CABPK.

Reply:
updated